### ISSUE TITLE
Don't install into ~/fisherman by default

### ISSUE THREAD

**Issue by szhu** [fisherman/fisherman#59]
The tidiness of my home folder is important to me. Please don't intrude!

<img width="174" src="https://cloud.githubusercontent.com/assets/1570168/12415874/f7befe0c-be52-11e5-95d2-314a29528a8d.png">

Perhaps put it under `~/.local/fisherman` or `~/.local/opt/fisherman`?

**Comment by bucaran** [comment]
:smile: 

Actually, Fisherman does _not_ install to `$HOME` by default, it installs to $PWD by default. You can change this location by moving to your desired path and then install, but I reckon this is not specified anywhere.

The original method to install was by `git clone; and make` so it was **obvious** Fisherman would install to the current directory, but since the new advertised method just runs a script, I think we should install Fisherman to `~/.local/fisherman` by **default**.

What do you think?

**Comment by szhu** [comment]
Yes, good idea. fisherman is a global command (you can type `fisher` in any cd). This means that the install location should be the same regardless of the cd as well.

**Comment by bucaran** [comment]
Starting from 0.5.0, `install.fisherman.sh` will retrieve a more sophisticated install script.

If you have any other setup feature requests or ideas, now it would be a great time to let me know :smile:

**Comment by etu** [comment]
Just putting a note here, but default for `$XDG_DATA_HOME` is `$HOME/.local/share/` and I'd say that `$HOME/.local/share/fisherman` seems reasonable to follow XDG.

**Comment by daenney** [comment]
Link to the full spec: http://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html

**Comment by bucaran** [comment]
@etu @daenney Thanks folks! That's exactly the plan. I will be shipping this as soon as I am done with 0.5.0.

See also [here](https://github.com/fisherman/fisherman/pull/60).

**Comment by fibo** [comment]
I agree to install it in an hidden folder, `$HOME/.local/share/fisherman` looks good, but consider not all the people will have `$XDG_DATA_HOME` set. I discovered this variable reading this issue.

My use case was: ok, let's trust the author and install it from install.fisherman.sh, and I think that it should default to an hidden folder even if no variable was found.

**Comment by bucaran** [comment]
Yep.

If they don't have `XDG_DATA_HOME`, which is very common (I don't have it set), we install to `$HOME/.local/share/fisherman` by default, else, we install to `$XDG_DATA_HOME/fisherman`.

**Comment by bucaran** [comment]
@szhu @fibo @simnalamburt @daenney @etu Hi everyone, the new [Fisherman installer](https://github.com/fisherman/installer) is up and running.

:anchor:

**Comment by fibo** [comment]
I got this error, after removing fisher

```
gcasati@Beintoo ~> fisher install shellder
Installing >> shellder
source: Error encountered while sourcing file '/Users/gcasati/fisherman/completions/fisher.fish':
source: No such file or directory
Aye! 1 plugin/s installed in 1s
```

**Comment by bucaran** [comment]
@fibo 

0.5.0 was a huge update, and this is probably due to the changes in the internal API. For example, there is now a new $fisher_file and $fisher_prompt variables used to check what plugins are currently installed and I think that's the issue. I recommend a clean install from zero.

If you still have the same issue by all means create a new issue for that :sweat_smile:

**Comment by fibo** [comment]
no prob,

well, now if I start typing `fisher install b` everything blocks, maybe there is some issue with autocompletion?

**Comment by bucaran** [comment]
That's odd, autocompletion should be working just fine.

Can you confirm that `$fisher_file` exists?

**Comment by pickfire** [comment]
It seems like I found fisherman cache in `~/.config/fisherman/cache`, is the correct? I thought it should be `~/.cache/fisherman`?

**Comment by bucaran** [comment]
That's correct, the `cache` is used internally by Fisherman so the used should not be able to delete it. Perhaps another name would be less confusing?

**Comment by etu** [comment]
@bucaran If it's used internaly and not supposed to be altered by the user. The proper location would be `$XDG_DATA_DIR` or `$XDG_CACHE_DIR` depending on the nature of the data.

**Comment by bucaran** [comment]
So, this "cache" is basically where repos get downloaded to. When you install a plugin it is downloaded to this "cache" for the first time only. When you uninstall a plugin it only gets removed from there if you use `--force`.

Since Fisherman uses a "flat" tree by design, we can't simply add the paths to the downloaded repos to `$fish_function_path` which is what I did originally with Wahoo and Oh My Fish!. 

Why? Because it gets slow as you need to juggle with more and more paths not to mention the PITA of mutating that array as the user installs / uninstalls plugins.

In Fisherman there is only one path added to the `$fish_function_path`, think of it as a huge Megazord plugin that is loaded only once, which is why Fisherman is fast.

Now, I also use the cache to store a copy of the index (plugin database), and your fishfile. Both the cache `$fisher_cahe` and the fishfile `$fisher_file` locations are configurable.

I don't see the benefit of moving the cache outside `$fisher_config` by default. But if you think I am wrong please give me your arguments, I am also learning here.

**Comment by etu** [comment]
So, I'll just link this blog post: https://ploum.net/207-modify-your-application-to-use-xdg-folders/

At the moment, I'm not very deep into the inner functions of how fisherman works, or how the tools are built. Or even exactly what they do when you run different things.

At the moment, I'm looking at this from a `user`-perspective knowing how XDG is supposed to split up applications.

If I remove `~/.cache/` and start a new terminal with a fresh shell. Will it break? Will everything just work as before, but maybe slower on first load?

**Comment by bucaran** [comment]
@etu I am afraid that break, it will.

If the naming is an issue, how do you feel about `stage`? Because that would not be so distant from the truth either.

**Comment by etu** [comment]
@bucaran The naming isn't the issue, I think. Just the location, I believe. The article explains quite well the user-reaction on removal of each directory.

**Comment by bucaran** [comment]
@etu Thanks for linking that up. It looks like a friendly read. I'll go over it and come back to this later.
